server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/cs/deployment/AbstractCacheStoreExtensionProcessor.java

```diff
@@ -0,0 +1,93 @@
+      Module module = deploymentUnit.getAttachment(Attachments.MODULE);
+      ServicesAttachment servicesAttachment = deploymentUnit.getAttachment(Attachments.SERVICES);
+      if (module != null && servicesAttachment != null)
+         addServices(ctx, servicesAttachment, module);
+   }
+
+   @Override
+   public void undeploy(DeploymentUnit deploymentUnit) {
+      // Deploy only adds services, so no need to do anything here
+      // since these services are automatically removed.
+   }
+
+   private void addServices(DeploymentPhaseContext ctx, ServicesAttachment servicesAttachment, Module module) {
+      Class<T> serviceClass = getServiceClass();
+      List<String> implementationNames = servicesAttachment.getServiceImplementations(serviceClass.getName());
+      ModuleClassLoader classLoader = module.getClassLoader();
+      for (String serviceClassName : implementationNames) {
+         try {
+            Class<? extends T> clazz = classLoader.loadClass(serviceClassName).asSubclass(serviceClass);
+            Constructor<? extends T> ctor = clazz.getConstructor();
+            T instance = ctor.newInstance();
+            installService(ctx, serviceClassName, instance);
+         } catch (Exception e) {
+            throw new IllegalStateException("Could not instantiate class " + serviceClassName, e);
+         }
+      }
+   }
+
+   public final void installService(DeploymentPhaseContext ctx, String implementationClassName, T instance) {
+      AbstractExtensionManagerService<T> service = createService(implementationClassName, instance);
+      ServiceName extensionServiceName = ServiceName.JBOSS.append(service.getServiceTypeName(), implementationClassName.replaceAll("\\.", "_"));
+      logger.info("Registering Deployed Cache Store service for store " + implementationClassName);
+      ServiceBuilder<T> serviceBuilder = ctx.getServiceTarget().addService(extensionServiceName, service);
+      serviceBuilder.setInitialMode(ServiceController.Mode.ACTIVE);
+      serviceBuilder.addDependency(DeployedCacheStoreFactoryService.SERVICE_NAME, DeployedCacheStoreFactory.class, service.getDeployedCacheStoreFactory());
+      serviceBuilder.install();
+   }
+
+   public abstract Class<T> getServiceClass();
+
+   public abstract AbstractExtensionManagerService<T> createService(String serviceName, T instance);
+
+   protected static abstract class AbstractExtensionManagerService<T> implements Service<T> {
+
+      protected final T extension;
+      protected final String className;
+      protected InjectedValue<DeployedCacheStoreFactory> deployedCacheStoreFactory = new InjectedValue<>();
+
+      protected AbstractExtensionManagerService(String className, T extension) {
+         this.extension = extension;
+         this.className = className;
+      }
+
+      @Override
+      public void start(StartContext context) {
+         logger.debugf("Started AdvancedCacheLoader service with name = %s", className + " " + deployedCacheStoreFactory);
```

Comment:
To be honest - I don't know. It is placed in `spi` package, so I would consider it as a part of public API.